fedbiomed/common/secagg_manager.py

```diff
@@ -113,30 +122,60 @@ def _add_generic(self, secagg_id: str, parties: List[str], specific: dict):
             FedbiomedSecaggError: failed to insert in database
             FedbiomedSecaggError: an entry already exists for `secagg_id` in the table
         """
-        if self._get_generic(secagg_id) is not None:
-            errmess = f'{ErrorNumbers.FB623.value}: error adding element in table "{self._table}": ' \
-                      f' an entry already exists for secagg_id={secagg_id}'
-            logger.error(errmess)
-            raise FedbiomedSecaggError(errmess)
-
+        self._check_existing_entry_in_db(secagg_id)
         specific.update({
             'secagg_version': str(__secagg_element_version__),
             'secagg_id': secagg_id,
-            'parties': parties
+            'parties': parties,
+            'secagg_scheme': secagg_scheme.value,
+            'secagg_elem': secagg_elem.value
         })
+
         try:
             self._table.insert(specific)
         except Exception as e:
             errmess = f'{ErrorNumbers.FB623.value}: failed adding an entry in table "{self._table}" ' \
-                      f'for secagg element secagg_id={secagg_id} with error: {e}'
+                      f'for secagg element={secagg_elem.name} secagg_id={secagg_id} with error: {e}'
+            logger.error(errmess)
+            raise FedbiomedSecaggError(errmess)
+
+    def _check_existing_entry_in_db(self, secagg_id: str):
+        if self._get_generic(secagg_id) is not None:
+            errmess = f'{ErrorNumbers.FB623.value}: error adding element in table "{self._table}": ' \
+                      f' an entry already exists for secagg_id={secagg_id}'
+            logger.error(errmess)
+            raise FedbiomedSecaggError(errmess)
+
+    def _raise_error_incompatible_retrieved_entry(self,
+                                                  entry: Union[None, Dict],
+                                                  experiment_id: str,
+                                                  secagg_id: str,
+                                                  component: SecaggElementTypes):
+        if entry is not None and entry['experiment_id'] != experiment_id:
```

Comment:
Thank you @ybouilla for this PR !

Here shouldn't we also test  that the `entry['secagg_elem']` matches the `component.value` ?

Reply:
> Here shouldn't we also test that the entry['secagg_elem'] matches the component.value ?

Yes this sounds like a good idea!

> Plus we should also call _raise_error_incompatible_retrieved_entry() from SecaggBiprimeManager.get() around line 330 ...

For that I am not sure. In the previous version there is no error raised, so I decided to leave it as it is. Is it something we should change?

Reply:
>>  Plus we should also call _raise_error_incompatible_retrieved_entry() from SecaggBiprimeManager.get() around line 330 ...
> For that I am not sure. In the previous version there is no error raised, so I decided to leave it as it is. Is it something we should change?

Thank you @ybouilla for your answer.

Sorry, I really think this is not coherent.
Of course in the previous version we don't test that the entry['secagg_elem'] matches the component.value in `SecaggBiprimeManager.get()`. We also don't test it in `SecaggServkeyManager.get()`.
This is perfectly fine in both cases as in previous version the database table is **dedicated** to a **single** `secagg_element` type. So we cannot retrieve another type of element.

Now the database table is shared between all types of context element (servkey, biprime, DH) so when we retrieve an element we always need to ensure that the element is of the proper type ...

If it is not clear to you maybe we can discuss it this afternoon ?

Reply:
Oh ok Sorry I misunderstood: you are speaking about the check against `secagg_element`, correct?
I thought you were speaking about `experiment_id` (which is not checked in `SecaggBiprimeManager`). Actually ` _raise_error_incompatible_retrieved_entry` is first and foremost checking if `experiment_id`s match, but is unused in `SecaggBiprimeManager`. So I would create another method that checks correctness of `secagg_element`. 
Let me know if there is something I am missing or misunderstanding

Reply:
> you are speaking about the check against secagg_element, correct?

Yes, exactly.

> Actually  _raise_error_incompatible_retrieved_entry is first and foremost checking if experiment_ids match, but is unused in SecaggBiprimeManager. So I would create another method that checks correctness of secagg_element.

Yes, I agree with that.
We still don't want to check `experiment_id` for `SecaggBiprimeManager`

Reply:
> Plus we should also call _raise_error_incompatible_retrieved_entry() from SecaggBiprimeManager.get() around line 330 ...

Adding a test in `SecaggBiprimeManager.get()`